### PR TITLE
Grouped bar

### PR DESCRIPTION
**Pull request by bung87**
related issue https://github.com/eventbrite/britecharts/issues/119
## Description

screen shot 2017-05-28 at 11 29 15 pm
screen shot 2017-05-28 at 11 29 25 pm

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by martinmanzo** [comment]
Hi, any news on this pull request?

File: src/charts/grouped-bar.js [hunk 1]
```diff
@@ -0,0 +1,969 @@
...
+     *     .call(groupedBar);
+     *
+     */
+    return function module() {
+
+        let margin = {
+            top: 40,
+            right: 30,
+            bottom: 60,
+            left: 70
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: One indentation position please

File: src/charts/grouped-bar.js [hunk 2]
```diff
@@ -0,0 +1,969 @@
...
+
+        /**
+         * Parses dates and values into JS Date objects and numbers
+         * @param  {obj} data Raw data from JSON file
+         * @return {obj}      Parsed data with values and dates
+         */
+        function cleanData(data) {
+            return data.map((d) => {
+                    d.value = +d[valueLabel];
+                    d.stack = d[groupLabel];
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: Could we call this 'group' instead of stack? This could be confusing

File: webpack.config.js [hunk 3]
```diff
@@ -69,7 +69,8 @@ config = {
             'demo-sparkline': './demos/demo-sparkline.js',
             'demo-step': './demos/demo-step.js',
             'demo-brush': './demos/demo-brush.js',
-            'demo-kitchen-sink': './demos/demo-kitchen-sink.js'
+            'demo-kitchen-sink': './demos/demo-kitchen-sink.js',
+            'demo-grouped-bar': './demos/demo-grouped-bar.js'
```
Comment (Golodhros) [inline webpack.config.js]: Please add it to 'src/bundle.js' too. Thanks

File: src/charts/grouped-bar.js [hunk 4]
```diff
@@ -0,0 +1,969 @@
...
+         * @private
+         */
+        function buildAxis() {
+            if (!horizontal) {
+                xAxis = d3Axis.axisBottom(xScale)
+                yAxis = d3Axis.axisLeft(yScale)
+                    .ticks(numOfVerticalTicks, valueLabelFormat)
+            } else {
+                xAxis = d3Axis.axisBottom(xScale)
+                .ticks(numOfHorizontalTicks, valueLabelFormat);
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: One indentation space, thanks

File: src/charts/grouped-bar.js [hunk 5]
```diff
@@ -0,0 +1,969 @@
...
+                .append('g').classed('metadata-group', true);
+        }
+
+        /**
+         * Builds the stacked layers layout
+         * @return {D3Layout} Layout for drawing the chart
+         * @private
+         */
+        function buildLayers(){
+            let dataInitial = transformedData.map((item) => {
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: We could skip the creation of the temporal variable 'dataInitial' and assign the result directly into 'layers'

File: src/charts/grouped-bar.js [hunk 6]
```diff
@@ -0,0 +1,969 @@
...
+
+            if (isAnimated){
+                bars.style('opacity', 0.24)
+                    .transition()
+                    .delay((_, i) => animationDelays[i])
+                    .duration(animationDuration)
+                    .ease(ease)
+                    .tween('attr.width', function(d ){
+                        let node = d3Selection.select(this),
+                        i = d3Interpolate.interpolateRound(0,xScale( getValue(d) )),
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: Indentation

File: src/charts/grouped-bar.js [hunk 7]
```diff
@@ -0,0 +1,969 @@
...
+                            .attr('height', yScale2.bandwidth())
+                            .attr('fill', ((data) => categoryColorMap[data.stack]));
+
+            if (isAnimated){
+                bars.style('opacity', 0.24)
+                    .transition()
+                    .delay((_, i) => animationDelays[i])
+                    .duration(animationDuration)
+                    .ease(ease)
+                    .tween('attr.width', function(d ){
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: I would love to see this tweening function extracted into a named function

File: src/charts/grouped-bar.js [hunk 8]
```diff
@@ -0,0 +1,969 @@
...
+                        let node = d3Selection.select(this),
+                        i = d3Interpolate.interpolateRound(0,xScale( getValue(d) )),
+                        j = d3Interpolate.interpolateNumber(0,1)
+                        ;
+
+                        return function(t){
+                            node.attr('width',  i(t) );
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: I think you can chain this two, right?

File: src/charts/grouped-bar.js [hunk 9]
```diff
@@ -0,0 +1,969 @@
...
+                .on('mousemove', function(d) {
+                    let data = assign({},this.__data__),[x,y] =  d3Selection.mouse(this);
+
+                    data.values = this.parentNode.__data__.values;
+                    dispatcher.call('customMouseMove', this, data,categoryColorMap,x,y);
+                })
+                .on('mouseout', function(d) {
+                    dispatcher.call('customMouseOut', this);
+                    d3Selection.select(this).attr('fill', () => categoryColorMap[d.stack])
+                });
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: This looks the same as the previous method, could we remove some duplication?

File: .eslintrc.json [hunk 10]
```diff
@@ -48,5 +48,8 @@
             "let": "always",
             "const": "never"
         }]
+    }, 
+       "parserOptions": {
```
Comment (Golodhros) [inline .eslintrc.json]: Nit: indentation

File: src/charts/grouped-bar.js [hunk 11]
```diff
@@ -0,0 +1,988 @@
...
+            nameLabel = 'name',
+            valueLabel = 'value',
+            groupLabel = 'group',
+            nameLabelFormat,
+            valueLabelFormat = NUMBER_FORMAT,
+
+            // getters
+            getName = (data) => data[nameLabel],
+            getValue = (data) => data[valueLabel],
+            getGroup = (data) => data[groupLabel],
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: These three should be either:
  getName = ({name}) => name,
  getValue = ({value}) => value,
  ...

Or we should remove them

File: src/charts/grouped-bar.js [hunk 12]
```diff
@@ -0,0 +1,988 @@
...
+            });
+        }
+
+        /**
+         * Creates the x, y and color scales of the chart
+         * @private
+         */
+        function buildScales() {
+
+            let yMax = d3Array.max(data.map(function (d) {
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: getValue

File: src/charts/grouped-bar.js [hunk 13]
```diff
@@ -0,0 +1,988 @@
...
+         * Draws grid lines on the background of the chart
+         * @return void
+         */
+        function drawGridLines(xTicks, yTicks) {
+            if (grid === 'horizontal' || grid === 'full') {
+                svg.select('.grid-lines-group')
+                    .selectAll('line.horizontal-grid-line')
+                    .data(yScale.ticks(yTicks).slice(1))
+                    .enter()
+                    .append('line')
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: Here we want the 'append' to have 2 spaces of indentation, to mark where we are changing the selection

File: src/charts/grouped-bar.js [hunk 14]
```diff
@@ -0,0 +1,988 @@
...
+                    .attr('y1', (d) => yScale(d))
+                    .attr('y2', (d) => yScale(d));
+            }
+
+            if (grid === 'vertical' || grid === 'full') {
+                svg.select('.grid-lines-group')
+                    .selectAll('line.vertical-grid-line')
+                    .data(xScale.ticks(xTicks).slice(1))
+                    .enter()
+                    .append('line')
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: Same here

File: src/charts/grouped-bar.js [hunk 15]
```diff
@@ -0,0 +1,988 @@
...
+        /**
+         * Animation tween of horizontal bars
+         * @param  {obj} d data of bar
+         * @return {void}
+         */
+        function horizontalBarsTween(d) {
+            let node = d3Selection.select(this),
+                i = d3Interpolate.interpolateRound(0, xScale(getValue(d))),
+                j = d3Interpolate.interpolateNumber(0, 1)
+                ;
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: Nit: semicolon in previous line

File: src/charts/grouped-bar.js [hunk 16]
```diff
@@ -0,0 +1,988 @@
...
+         */
+        function verticalBarsTween(d) {
+            let node = d3Selection.select(this),
+                i = d3Interpolate.interpolateRound(0, chartHeight - yScale(getValue(d))),
+                y = d3Interpolate.interpolateRound(chartHeight, yScale(getValue(d))),
+                j = d3Interpolate.interpolateNumber(0, 1)
+                ;
+
+            return function (t) {
+                node.attr('y', y(t)).attr('height', i(t)).style('opacity', j(t));
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: Nit: let's put this in multiple lines

File: src/charts/grouped-bar.js [hunk 17]
```diff
@@ -0,0 +1,988 @@
...
+        function handleMouseMove() {
+            let mousePos = getMousePosition(this),
+                dataPoint = !horizontal ? getNearestDataPoint(mousePos) : getNearestDataPoint2(mousePos),
+                x,
+                y;
+
+            if (dataPoint) {
+                // Move verticalMarker to that datapoint
+                if (!horizontal) {
+                    x = xScale(dataPoint.key) + xScale2(dataPoint[groupLabel]), y = yScale(getValue(dataPoint));
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: I am a bit confused by that ',' after the x assignement

**Comment by qmn1711** [comment]
Really cool feature guys, any news on this pull request?

**Comment by Golodhros** [comment]
I just arrived from vacations. I will be tackling this PR ASAP!

**Comment by Golodhros** [comment]
Merged! Many thanks @bung87  and sorry for the delay!

**Pull request closed without merging**